File: app/src/main/java/ss/nscube/webshare/ServerService.java
package ss.nscube.webshare;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.app.TaskStackBuilder;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.os.Binder;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.viewpager.widget.PagerAdapter;

import ss.nscube.webshare.server.HTTPServer;

public class ServerService extends Service {
    private final IBinder binder = new LocalBinder();
    private HTTPServer server;
    private NotificationCompat.Builder notificationBuilder;
//    private NotificationCompat.Action pauseAction, continueAction;
//    public static final String PAUSE = "PAUSE";
//    public static final String CONTINUE = "CONTINUE";
    public static final String STOP_SERVER_ACTION = "STOP_SERVER";
//    public static final String PAUSE_SERVER_ACTION = "PAUSE_SERVER";
//    public static final String CONTINUE_SERVER_ACTION = "CONTINUE_SERVER";

    private NotificationManager notificationManager;

    @Override
    public void onCreate() {
        super.onCreate();
        ServerBroadCastReceiver.serverService = this;
        server = new HTTPServer(getAssets());
        notificationBuilder = new NotificationCompat.Builder(this, "channelId");
        notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
//        pauseAction = new NotificationCompat.Action(0, PAUSE, makePendingIntent(PAUSE_SERVER_ACTION));
//        continueAction = new NotificationCompat.Action(0, CONTINUE, makePendingIntent(CONTINUE_SERVER_ACTION));
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    private void setupNotification() {

        Notification notification = null;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {

            NotificationChannel chan = new NotificationChannel("channelId", "HTTP Server", NotificationManager.IMPORTANCE_MIN);
            chan.setLightColor(Color.BLUE);
            chan.setLockscreenVisibility(Notification.VISIBILITY_PRIVATE);
            assert notificationManager != null;
            notificationManager.createNotificationChannel(chan);
        }

        PendingIntent activityIntent = PendingIntent.getActivity(this, 0, new Intent(this, MainActivity.class), PendingIntent.FLAG_CANCEL_CURRENT);

        notification = notificationBuilder.setOngoing(true)
                .setDefaults(Notification.DEFAULT_LIGHTS | Notification.DEFAULT_SOUND)
                .setVibrate(new long[]{0L})
                .setSmallIcon(R.drawable.notification_icon)
                .setContentTitle("WebShare is running in background")
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setCategory(Notification.CATEGORY_SERVICE)
                .setContentIntent(activityIntent)
                .addAction(0, "STOP", makePendingIntent(STOP_SERVER_ACTION))
                .build();

        startForeground(1, notification);
    }

    public PendingIntent makePendingIntent(String name) {
        Intent intent = new Intent(this, ServerBroadCastReceiver.class);
        intent.setAction(name);
        return PendingIntent.getBroadcast(this, 0, intent, 0);
    }

    public HTTPServer getServer() {
        return server;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        stopForeground(true);
        return binder;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        if (server.getState() == HTTPServer.RUNNING || server.getState() == HTTPServer.PAUSED) {
            setupNotification();
        } else {
            stopForeground(true);
        }
        return super.onUnbind(intent);
    }

    public  class LocalBinder extends Binder {
        ServerService getServerService() {
            return ServerService.this;
        }
    }

    public static class ServerBroadCastReceiver extends BroadcastReceiver {
        static ServerService serverService;

        public ServerBroadCastReceiver() {
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            switch (intent.getAction()) {
                case ServerService.STOP_SERVER_ACTION:
                    serverService.getServer().stop();
                    serverService.stopForeground(true);
                    serverService.stopSelf();
                    break;
            }

        }
    }
}

File: app/src/main/java/ss/nscube/webshare/ServerService.kt
package ss.nscube.webshare

import android.app.*
import android.os.IBinder
import ss.nscube.webshare.server.HTTPServer
import android.content.Intent
import androidx.core.app.NotificationCompat
import ss.nscube.webshare.ui.MainActivity
import android.graphics.Color
import android.os.Binder
import android.os.Build
import ss.nscube.webshare.utils.log

class ServerService : Service() {
    private val binder: IBinder = LocalBinder()
    val channelId = "WebShareServer"
    val flag = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_CANCEL_CURRENT
    } else {
        PendingIntent.FLAG_CANCEL_CURRENT
    }

    override fun onCreate() {
        super.onCreate()
        log("SERVICE onCreate")
        setupNotification()
    }

    val server: HTTPServer
        get() {
            return (application as WebShareApp).server
        }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        log("SERVICE onStartCommand ${intent?.action} ${Thread.currentThread().name}")
        when(intent?.action) {
            STOP_SERVER_ACTION -> {
                server.stop()
            }
            else -> {
            }
        }
        return START_STICKY
    }

    private fun setupNotification() {
        val notificationBuilder = NotificationCompat.Builder(this, channelId)
        val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val chan =
                NotificationChannel(channelId, "HTTP Server", NotificationManager.IMPORTANCE_MIN)
            chan.lightColor = Color.BLUE
            chan.lockscreenVisibility = Notification.VISIBILITY_PRIVATE
            notificationManager.createNotificationChannel(chan)
        }

        val notification = notificationBuilder.setOngoing(true)
            .setDefaults(Notification.DEFAULT_LIGHTS or Notification.DEFAULT_SOUND)
            .setVibrate(longArrayOf(50L))
            .setSmallIcon(R.drawable.notification_icon)
            .setContentTitle("WebShare is running in the background")
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setCategory(Notification.CATEGORY_SERVICE)
            .setContentIntent(
                PendingIntent.getActivity(
                    this,
                    0,
                    Intent(this, MainActivity::class.java),
                    flag
                )
            )
            .addAction(0, "STOP", makePendingIntent(STOP_SERVER_ACTION))
            .build()
        startForeground(1, notification)
    }

    fun makePendingIntent(name: String?): PendingIntent {
        val stopSelf = Intent(this, ServerService::class.java);
        stopSelf.action = STOP_SERVER_ACTION;
        return PendingIntent.getService(this, 0, stopSelf, flag)
    }

    override fun onBind(intent: Intent): IBinder? {
        return binder
    }

    override fun onUnbind(intent: Intent): Boolean {
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        super.onDestroy()
        log("SERVICE onDestroy")
    }

    inner class LocalBinder : Binder() {
        val serverService: ServerService
            get() = this@ServerService
    }

    companion object {
        const val STOP_SERVER_ACTION = "STOP_SERVER"
    }
}